runelite-client/src/main/java/net/runelite/client/plugins/groundmarkers/GroundMarkerPlugin.java

```diff
@@ -0,0 +1,438 @@
+	{
+		return overlay;
+	}
+
+	protected void markTile(Point ptMouse)
+	{
+		LocalPoint localPoint = getTile(ptMouse);
+
+		if (localPoint == null)
+		{
+			return;
+		}
+
+		WorldPoint worldPoint;
+
+		if (client.isInInstancedRegion())
+		{
+			// get position in the scene
+			int sceneX = localPoint.getRegionX();
+			int sceneY = localPoint.getRegionY();
+
+			// get chunk from scene
+			int chunkX = sceneX / CHUNK_SIZE;
+			int chunkY = sceneY / CHUNK_SIZE;
+
+			// get the template chunk for the chunk
+			int[][][] instanceTemplateChunks = client.getInstanceTemplateChunks();
+			int templateChunk = instanceTemplateChunks[client.getPlane()][chunkX][chunkY];
+
+			int rotation = templateChunk >> 1 & 0x3;
+			int templateChunkY = (templateChunk >> 3 & 0x7FF) * CHUNK_SIZE;
+			int templateChunkX = (templateChunk >> 14 & 0x3FF) * CHUNK_SIZE;
+			int plane = templateChunk >> 24 & 0x3;
+
+			// calculate world point of the template
+			int x = templateChunkX + (sceneX & (CHUNK_SIZE - 1));
+			int y = templateChunkY + (sceneY & (CHUNK_SIZE - 1));
+
+			worldPoint = new WorldPoint(x, y, plane);
+			// rotate point back to 0, to match with template
+			worldPoint = rotate(worldPoint, rotation);
+		}
+		else
+		{
+			worldPoint = WorldPoint.fromLocal(client, localPoint);
+		}
+
+		int regionX = worldPoint.getX() >> 6;
+		int regionY = worldPoint.getY() >> 6;
+		int regionId = regionX << 8 | regionY;
+		GroundMarkerPoint point = new GroundMarkerPoint(regionId, worldPoint.getX() & 0x3f, worldPoint.getY() & 0x3f, client.getPlane());
+		log.debug("Updating point: {} - {}", point, worldPoint);
+
+		List<GroundMarkerPoint> points = new ArrayList<>(getPoints(regionId));
+		if (points.contains(point))
+		{
+			points.remove(point);
+		}
+		else
+		{
+			points.add(point);
+		}
+
+		savePoints(regionId, points);
+
+		loadPoints();
+	}
+
+	/**
+	 * Rotate the chunk containing the given point to rotation 0
+	 *
+	 * @param point
+	 * @param rotation
+	 * @return
+	 */
+	private WorldPoint rotate(WorldPoint point, int rotation)
+	{
+		int chunkX = point.getX() & ~(CHUNK_SIZE - 1);
+		int chunkY = point.getY() & ~(CHUNK_SIZE - 1);
+		int x = point.getX() & (CHUNK_SIZE - 1);
+		int y = point.getY() & (CHUNK_SIZE - 1);
+		switch (rotation)
+		{
+			case 1:
+				return new WorldPoint(chunkX + ((CHUNK_SIZE - 1) - y), chunkY + x, point.getPlane());
+			case 2:
+				return new WorldPoint(chunkX + ((CHUNK_SIZE - 1) - x), chunkY + ((CHUNK_SIZE - 1) - y), point.getPlane());
+			case 3:
+				return new WorldPoint(chunkX + y, chunkY + ((CHUNK_SIZE - 1) - x), point.getPlane());
+		}
+		return point;
+	}
+
+	/**
+	 * Find the tile which the given mouse point is in
+	 * @param mouse
+	 * @return
+	 */
+	private LocalPoint getTile(Point mouse)
+	{
+		Region region = client.getRegion();
+		Tile[][][] tiles = region.getTiles();
+		int z = client.getPlane();
+
+		java.awt.Point gameMouse = new java.awt.Point(mouse.getX(), mouse.getY());
+
+		for (int x = 0; x < REGION_SIZE; ++x)
```

Comment:
Use RegionTileManager.forEachTile

Reply:
If I did that I'd have to return from inside of a lambda, which in Java is more trouble than its worth really..